### PR TITLE
Add `Str::isEmpty()`

### PR DESCRIPTION
**Pull request by TheSyscall**
Adds a static `isEmpty()` method to the `Str` class.

`Str::isEmpty()` is meant for a check if a string is null, or exactly '' after trimming whitespace.

~Add some static methods to the Str class.~

~- `endsWith` as the counterpart to the existing `startsWith`~
~- `contains` to have a case independent version of `str_contains`~
~- `containsAny` and `containsAll` to quickly compare against an array of strings~
~- `isEmpty` to reliably check for empty strings~

Moved to #78

### CODE REVIEW

File: tests/StrTest.php [hunk 1]
```diff
@@ -47,6 +47,152 @@ public function testStartsWithReturnsFalseIfStringDoesNotStartWithTheSpecifiedSu
...
+    }
+
+    public function testIsEmptyReturnsFalseForStringWithContentAndSurroundingWhitespace()
+    {
+        $this->assertFalse(Str::isEmpty('  Warning  '));
+    }
+
+    public function testIsEmptyReturnsTrueForStringWithCustomCharacters()
+    {
+        $this->assertTrue(Str::isEmpty('---', '-'));
```
Comment (Al2Klimov) [inline tests/StrTest.php]: Please add test cases for the stuff where empty() breaks in contrast. E.g: '0'

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
LGTM

File: tests/StrTest.php (hunk 1 above)
Comment (TheSyscall) [reply tests/StrTest.php]: Added explicit test for `isEmpty('0')`

**Review by Al2Klimov — DISMISSED** [review]
Please hand over to the next reviewer ASAP!

**Comment by yhabteab** [comment]
Hi, I don't think, I can give any meaningful reviews on this, thus, it would probably be better to request a review from @nilmerg instead, if Eric is not available.

File: src/Str.php [hunk 2]
```diff
@@ -46,6 +46,84 @@ public static function startsWith(?string $subject, string $start, bool $caseSen
...
+     * @param string $end
+     * @param bool $caseSensitive
+     *
+     * @return bool
+     */
+    public static function endsWith(?string $subject, string $end, bool $caseSensitive = true): bool
+    {
+        $subject ??= '';
+        if (! $caseSensitive) {
+            return strncasecmp($subject, $end, strlen($end)) === 0;
```
Comment (lippserd) [inline src/Str.php]: This is a prefix check but it should be a suffix check. Tests don’t exercise the contract properly. You always pass true, which is the default.

File: src/Str.php [hunk 3]
```diff
@@ -46,6 +46,84 @@ public static function startsWith(?string $subject, string $start, bool $caseSen
...
+        }
+
+        return str_contains($subject, $needle);
+    }
+
+    /**
+     * Check if the given string contains any of the specified substrings
+     *
+     * @param ?string $haystack
+     * @param string[] $needles
```
Comment (lippserd) [inline src/Str.php]: Use iterable to support any string iterator/generator.

File: src/Str.php [hunk 4]
```diff
@@ -46,6 +46,84 @@ public static function startsWith(?string $subject, string $start, bool $caseSen
...
+        }
+
+        return false;
+    }
+
+    /**
+     * Check if the given string contains all the specified substrings
+     *
+     * @param ?string $haystack
+     * @param string[] $needles
```
Comment (lippserd) [inline src/Str.php]: Same iterable consideration as above.

File: src/Str.php [hunk 5]
```diff
@@ -90,4 +168,19 @@ public static function trimSplit(?string $subject, string $delimiter = ',', ?int
...
         return array_map('trim', $exploded);
     }
+
+    /**
+     * Check if the given string is empty
+     *
+     * Null is considered empty and strings are trimmed before checking.
+     *
+     * @param ?string $subject
+     * @param string $characters
```
Comment (lippserd) [inline src/Str.php]: isEmpty should not expose implementation details. Its contract is that it considers null or whitespace only strings as empty. That’s it. Calling trim must be replaced anyway if we want this function to also work with UTF-8 inputs.

File: src/Str.php [hunk 6]
```diff
@@ -46,6 +46,84 @@ public static function startsWith(?string $subject, string $start, bool $caseSen
...
+     *
+     * @param ?string $subject
+     * @param string $needle
+     * @param bool $caseSensitive
+     *
+     * @return bool
+     */
+    public static function contains(?string $subject, string $needle, bool $caseSensitive = true): bool
+    {
+        $subject ??= '';
```
Comment (lippserd) [inline src/Str.php]: Exit early instead.

File: src/Str.php [hunk 7]
```diff
@@ -46,6 +46,84 @@ public static function startsWith(?string $subject, string $start, bool $caseSen
...
+     *
+     * @param ?string $subject
+     * @param string $end
+     * @param bool $caseSensitive
+     *
+     * @return bool
+     */
+    public static function endsWith(?string $subject, string $end, bool $caseSensitive = true): bool
+    {
+        $subject ??= '';
```
Comment (lippserd) [inline src/Str.php]: Exit early instead.

**Review by lippserd — CHANGES_REQUESTED** [review]
One more thing to consider: This caseSensitive parameter construct is something that I would do differently nowadays. Passing false without using a named argument requires the reader/caller to read the docs. It unnecessarily bloats the contract. And it requires two different implementation blocks in each function. Consider introducing …Insensitive() variants instead.

**Comment by TheSyscall** [comment]
> Consider introducing …Insensitive() variants instead.

That would introduce a lot of extra methods like `containsAnyInsensitive()`, moreover I can't change `startsWith()` leading to different behavior between `startsWith()` and `endsWith()`. And deprecating a perfectly fine function seems odd.

**Comment by lippserd** [comment]
> > Consider introducing …Insensitive() variants instead.
> 
> That would introduce a lot of extra methods like `containsAnyInsensitive()`, moreover I can't change `startsWith()` leading to different behavior between `startsWith()` and `endsWith()`. And deprecating a perfectly fine function seems odd.

Sure it would. On the other hand our lib would introduce new functions with a flawed contract (In addition to my points stated above, the bool should have been caseInsensitive: false). Anyway, how often do we really need the case insensitive variants? Do we need all the new introduced functions right away? Consider reducing/splitting the PR to what we actually need right now.

File: src/Str.php (hunk 5 above)
Comment (TheSyscall) [reply src/Str.php]: [mb_trim](https://www.php.net/manual/en/function.mb-trim.php) is not available for ipl-stdlib until we require php >= 8.4. I talked with @nilmerg, and checked for potential problems with only having ascii support for now.

It seems like we might have some possible false negatives but never any false positives (for a properly encoded utf-8 string at least).

File: src/Str.php (hunk 2 above)
Comment (TheSyscall) [reply src/Str.php]: Fixed. Moved to #78

File: src/Str.php (hunk 3 above)
Comment (TheSyscall) [reply src/Str.php]: Changed. Moved to #78

File: src/Str.php (hunk 4 above)
Comment (TheSyscall) [reply src/Str.php]: Moved to #78

File: src/Str.php (hunk 6 above)
Comment (TheSyscall) [reply src/Str.php]: Early return will break `contains('', '')` which is probably unexpected.

Moved to #78

File: src/Str.php (hunk 7 above)
Comment (TheSyscall) [reply src/Str.php]: This is a parity change with `Str::startsWith()`

Moved to #78

**Review by Al2Klimov — DISMISSED** [review]

File: src/Str.php (hunk 5 above)
Comment (TheSyscall) [reply src/Str.php]: Talking with @lippserd we decided to use a regex match to handle utf8 characters and also include visually empty characters in this method.

**Review by lippserd — APPROVED** [review]
